Написал UI тесты для сервиса Яндекс.Самокат

CRASH3000 opened the pull request:
Добавлены автотесты для учебного сервиса «Яндекс.Самокат». 
Тесты охватывают следующие сценарии:

### Тестирование выпадающего списка в разделе "Вопросы о важном":
* Проверяется, что при нажатии на вопрос открывается соответствующий текст.
* Написаны отдельные тесты для каждого вопроса.
* Проверка работы элементов в Google Chrome и Firefox.

### Тестирование флоу заказа:
* Проверяется весь процесс создания заказа с двумя наборами данных.
* Осуществляется проверка с двух точек входа: через кнопку "Заказать" в шапке страницы и на главной странице.
* Включает проверки всплывающего окна с подтверждением успешного создания заказа, а также кликов по логотипам «Самокат» и «Яндекс».
* Проверка работы элементов в Google Chrome и Firefox.

В рамках проекта реализовал:
* Модель Page Object для улучшения структуры кода и его поддержки.
* Использовал параметризацию для тестирования различных данных и точек входа (Без необходимости дублировать код теста).
* Интегрирован декораторы Allure для генерации отчетов о тестировании.
* Добавил файл README.md с инструкциями по установке, запуску тестов и генерации отчетов.

Структура проекта:
* Ознакомьтесь с тестами можно в папке tests.
* Проверьте структуру Page Object в папке pages и локаторы в папке locators.
* В папке data хранятся тестовые данные 
* Параметры запуска веб-драйвер можно найти в файле conftest.py

✅ Все тесты проверены и успешно проходят в браузерах Chrome и Firefox.

![Скрин Отчета3](https://github.com/user-attachments/assets/1c345018-0ba7-4af8-9cd5-0f24cd5feb5a)

![Скрин Отчета2](https://github.com/user-attachments/assets/37fb42be-c7c1-4a72-b8d7-de9aabae2e70)

![Скрин Отчета](https://github.com/user-attachments/assets/a6f22dad-3ff6-47ac-9985-9c04cf9944e5)

CRASH3000 committed:
Написал UI тесты для сервиса Яндекс.Самокат

- Реализованы сценарии тестирования для раздела "Вопросы о важном" и "Флоу заказа" с использованием Selenium WebDriver.
- Применена модель Page Object для улучшения структуры и поддержки кода.
- Добавлены параметризованные тесты для разных наборов данных и точек входа.
- Настроена интеграция с Allure для создания подробных отчетов о тестировании.
- Добавлен файл README.md с обзором проекта, инструкциями по установке и запуску, а также описанием тестов.

k-asadova commented on locators/main_page_locators.py:
```diff
@@ -0,0 +1,25 @@
+class MainPageLocators:
+
+    # Локаторы от блока "Вопросы о важном"
+    FAQ_SECTION = "//div[@class='Home_FAQ__3uVm4']"
```
Можно лучше: потенциально нестабильный локатор. Вот эта цифро-буквенная часть класса в конце строки Home_FAQ__**3uVm4** сгенерирована автоматически при деплое. При последующих апдейтах локатор может сломаться. Нужно использовать contains и только основное имя класса

CRASH3000 replied:
✅Просмотрел все локаторы и отредактировал проблемные.

k-asadova commented on tests/test_accordion.py:
```diff
@@ -0,0 +1,146 @@
+        assert (
+            answer_text == accordion_data["items"][0]["answer"]
+        ), f"Expected '{accordion_data['items'][0]['answer']}', but got '{answer_text}'"
+        time.sleep(3)
```
sleep - антипаттерн, лучше не использовать, вместо него - WebDriverWait

CRASH3000 replied:
✅ Удалил. Использовал sleep для проверки и отладки теста.

k-asadova commented on tests/test_accordion.py:
```diff
@@ -0,0 +1,146 @@
+        self.scroll_to_faq_section()
+        self.page.click_question(MainPageLocators.QUESTION_0)
+        answer_text = self.page.get_answer_text(MainPageLocators.ANSWER_0)
+        assert (
```
Нужно поправить: сейчас 8 тестов с вопросами-ответами содержат одну и ту же логику, просто проверяют разные локаторы. В таких случаях мы используем параметризацию

CRASH3000 replied:
✅ Переписал код теста для проверки элемента аккордеон. Применил параметризацию.

k-asadova commented on tests/test_create_order.py:
```diff
@@ -0,0 +1,208 @@
+@pytest.mark.usefixtures("driver")
+class TestOrder:
+
+    def setup_method(self, method):
+        self.main_page = MainPage(self.driver)
+        self.order_page = OrderPage(self.driver)
+        self.main_page.open()
+        self.wait = WebDriverWait(self.driver, 10)
+
+    @allure.step("Клик на кнопку 'Заказать' в шапке сайта")
+    def click_order_button_in_header(self):
+        self.main_page.click_order_button_in_header()
+        self.wait.until(EC.url_to_be(Urls.ORDER_URL))
+
+    @allure.step("Клик на кнопку 'Заказать' на главной странице")
+    def click_order_button_in_main_page(self):
+        self.main_page.close_cookie_message()
+        self.main_page.scroll_to_order_button()
+        self.main_page.click_order_button()
+        self.wait.until(EC.url_to_be(Urls.ORDER_URL))
```
Нужно поправить: все степы (логика) должны находиться в пейджах. В файлах с тестами - только тесты

CRASH3000 replied:
✅ Переписал код. Теперь в файле с тестами только тесты.

k-asadova commented on tests/test_create_order.py:
```diff
@@ -0,0 +1,208 @@
+    def test_order_form(self, data, click_order_button_method, confirm_order, request):
+
+        # Словари для сопоставления параметров с текстовыми значениями для названия теста в Allure
+        click_order_button_method_mapping = {
+            "click_order_button_in_header": 'через кнопку "Заказать" в шапке сайта',
+            "click_order_button_in_main_page": 'через кнопку "Заказать" на главной странице',
+        }
+
+        confirm_order_mapping = {
+            "yes": "с оформлением заказа",
+            "no": "с отменой заказа",
+        }
```
Нужно поправить: выносим тестовые данные за пределы тестового метода

CRASH3000 replied:
✅ Перенес логику для создания динамичных заголовков тестов в Allure в отдельный файл.

CRASH3000 committed:
Отработал комментарии ревьювера и внес правки:

Что нового?
- Создал отдельный файл allure_titles.py для хранения логики генерации заголовков и описаний текста для тестов в Allure с учетом дополнительных параметров, таких как номер набора данных, в рамках какого браузера был выполнен тест.
- Добавлена параметризация в тесте test_accordion.py

Что изменил:
- Отредактировал потенциально нестабильные локаторы для всех тестов.
- Переписал код теста для проверки элемента аккордеон. Применил параметризацию.
- Перенес из тестов степы (логика) в пейджи. Теперь в тестах только тесты и декораторы Allure.
- Внес небольшие правки в логику тестов.
- Переписал декораторы Allure в тестах.

Что удалил:
- Удалил однотипные тесты в test_accordion.py.
- Убрал из кода тестов sleep.
- Почистил кэш Allure от старых запусков тестов.
